README
NAME
    Git::Helpers - Shortcuts for common Git commands

VERSION
    version 0.000004

SYNOPSIS
        use Git::Helpers qw( checkout_root );
        my $root = checkout_root();

  checkout_root( $dir )
    Gives you the root level of the git checkout which you are currently in.
    Optionally accepts a directory parameter. If you provide the directory
    parameter, "checkout_root" will temporarily "chdir" to this directory
    and find the top level of the repository.

    This method will throw an exception if it cannot find a git repository
    at the directory provided.

AUTHOR
    Olaf Alders <olaf@wundercounter.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2015 by Olaf Alders.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Git-Helpers

{{$NEXT}}

1.000001  2021-11-02 01:31:05Z
    - Bump Term::Choose to latest version (1.743) (GH#18) (Olaf Alders)
    - Update use of Term::Choose in delete-git-branches (GH#16) (kuerbis)
    - Bump MetaCPAN::Client to latest version (2.029000) (GH#19) (Olaf Alders)
    - Bump minimum Perl to 5.12.0 (GH#19) (Olaf Alders)

1.000000  2021-02-09 20:39:56Z
    - Drop support for travis-ci.org (GH#15) (Olaf Alders)

0.000021  2021-02-08 22:59:26Z
    - Open correct URL for repositories with git URLs (GH#14) (Olaf Alders)

0.000020  2019-09-30 18:38:39Z
    - Remove dependency on say module (GH#12) (Olaf Alders)

0.000019  2019-09-25 16:31:33Z
    - Add --echo --version and --help to gh-open (GH#11) (Olaf Alders)

0.000018  2019-02-08 19:16:21Z
    - Add ignored_files() function (GH#10) (Olaf Alders)

0.000017  2018-12-18 21:55:01Z
    - https_remote_url() now returns earlier if it will fail. (GH#9) (Olaf Alders)

0.000016  2018-12-17 20:46:07Z
    - Add is_inside_work_tree() function. (GH#8) (Olaf Alders)

0.000015  2018-04-13 22:49:10Z
    - Use Term::Choose for a more intuitive UI

0.000014  2018-03-23 15:26:10Z
    - Fall back to config when get-url doesn't work (GH#7) (Olaf Alders)
    - Use #!perl shebang in scripts (GH#6) (Olaf Alders)

0.000013  2017-09-06 17:46:13Z
    - Really add delete-git-branches script

0.000012  2017-09-06 14:51:30Z
    - Add delete-git-branches script

0.000011  2017-05-31 10:11:30-04:00 America/Toronto
    - Capture STDERR in `checkout_root() (Chris Weyl)

0.000010  2017-04-27 13:55:03-04:00 America/Toronto
    - Adds cpan-repo script

0.000009  2017-01-12 16:28:17-05:00 America/Toronto
    - Fixes copy/paste typ on travis-open docs.

0.000008  2016-09-12 18:26:01-04:00 America/Toronto
    - Fix broken test.

0.000007  2016-09-09 17:33:58-04:00 America/Toronto
    - Adds current_branch_name() function
    - Adds gh-open -b arg, which opens the URL for your current branch (Kevin
      Phair)

0.000006  2016-08-12 00:02:57-04:00 America/Toronto
    - Adds URI::git to deps.
    - skip remote_url() tests if Git version less than 2.7

0.000005  2016-07-27 22:40:25-04:00 America/Toronto
    - Adds a remote_url() helper.
    - Adds a travis_url() helper.

0.000004  2016-04-11 21:57:50-04:00 America/Toronto
    - Replace has_git by test_requires_git (Philippe Bruhat (BooK))

0.000003  2015-04-20 04:35:06+02:00 Europe/Berlin
    - Fixes use of Test::Git

0.000002  2015-04-19 09:29:34+02:00 Europe/Berlin
    - Requires the (currently) latest version of Test::Git

0.000001  2015-04-18 15:00:31+02:00 Europe/Berlin
    - Initial release.


lib/Git/Helpers.pm
use strict;
use warnings;

package Git::Helpers;
our $VERSION = '1.000002';

use Carp qw( croak );
use Capture::Tiny qw( capture_stderr );
use File::pushd qw( pushd );
use Git::Sub;
use Sub::Exporter -setup => {
    exports => [
        'checkout_root',
        'current_branch_name',
        'https_remote_url',
        'ignored_files',
        'is_inside_work_tree',
        'remote_url',
    ]
};
use Try::Tiny qw( catch try );
use URI ();
use URI::Heuristic qw( uf_uristr );
use URI::git ();

sub checkout_root {
    my $dir = shift;

    my ( $new_dir, $root );
    $new_dir = pushd($dir) if $dir;
    $dir ||= '.';

    # the exception thrown by rev-parse when we're not in a repo is not
    # very helpful (see following), so just ditch it in favor of the error
    # output.
    #   "rev_parse error 128 at /home/maxmind/perl5/lib/perl5/Git/Helpers.pm line 30"

    my $stderr = capture_stderr {
        try { $root = scalar git::rev_parse qw(--show-toplevel) }
    };
    croak "Error in $dir: $stderr"
        if $stderr;

    return $root;
}

# Works as of 1.6.3:1
# http://stackoverflow.com/questions/1417957/show-just-the-current-branch-in-git
sub current_branch_name {
    return git::rev_parse( '--abbrev-ref', 'HEAD' );
}

sub https_remote_url {
    my $remote_url = remote_url(shift);
    return undef unless $remote_url;

    my $branch = shift;

    # remove trailing .git
    $remote_url =~ s{\.git\z}{};

    # remove 'git@' from git@github.com:username/repo.git
    $remote_url =~ s{\w*\@}{};

    # remove : from git@github.com:username/repo.git
    $remote_url =~ s{(\w):(\w)}{$1/$2};

    if ($branch) {
        $remote_url .= '/tree/' . current_branch_name();
    }

    my $uri = URI->new( uf_uristr($remote_url) );
    $uri->scheme('https');
    return $uri;
}

sub is_inside_work_tree {
    my $success;
    capture_stderr {
        try {
            $success = git::rev_parse('--is-inside-work-tree');
        };
    };
    return $success;
}

sub ignored_files {
    my $dir = shift || '.';
    my @success;
    my $stderr = capture_stderr {
        try {
            @success = git::ls_files(
                $dir, '--ignored', '--exclude-standard',
                '--others'
            );
        };
    };

    croak "Cannot find ignored files in dir: $stderr" if $stderr;

    return \@success || [];
}

sub remote_url {
    my $remote = shift || 'origin';
    my $url;
    my $stderr = capture_stderr {
        try {
            $url = git::remote( 'get-url', $remote );
        }

        catch {
            try {
                $url = git::config( '--get', "remote.$remote.url" );
            };
        }
    };

    return $url;
}

1;

#ABSTRACT: Shortcuts for common Git commands

=pod

=head1 SYNOPSIS

    use Git::Helpers qw(
        checkout_root
        current_branch_name
        https_remote_url
        is_inside_work_tree
        remote_url
    );

    my $dir              = '/path/to/folder/in/git/checkout';
    my $root             = checkout_root($dir);
    my $current_branch   = current_branch_name();
    my $https_remote_url = https_remote_url();
    my $inside_work_tree = is_inside_work_tree();
    my $remote_url       = remote_url('upstream');

=head2 checkout_root( $dir )

Gives you the root level of the git checkout which you are currently in.
Optionally accepts a directory parameter.  If you provide the directory
parameter, C<checkout_root> will temporarily C<chdir> to this directory and
find the top level of the repository.

This method will throw an exception if it cannot find a git repository at the
directory provided.

=head2 current_branch_name

Returns the name of the current branch.

=head2 https_remote_url( $remote_name, $use_current_branch )

This is a browser-friendly URL for the remote, fixed up in such a way that
GitHub (hopefully) doesn't need to redirect your URL.

Turns git@github.com:oalders/git-helpers.git into https://github.com/oalders/git-helpers

Turns https://github.com/oalders/git-helpers.git into https://github.com/oalders/git-helpers

Defaults to using C<origin> as the remote if none is supplied.

Defaults to master branch, but can also display current branch.

    my $current_branch_url = https_remote_url( 'origin', 1 );

=head2 ignored_files( $dir )

Returns an arrayref of files which exist in your checkout, but are ignored by
Git.  Optionally accepts a directory as an argument.  Defaults to ".".

Throws an exception if there has been an error running the command.

=head2 is_inside_work_tree

Returns C<true> if C<git rev-parse --is-inside-work-tree> returns C<true>.
Otherwise returns C<false>. This differs slightly from the behaviour of
C<--is-inside-work-tree> in real life, since it returns C<fatal> rather than
C<false> if run outside of a git repository.

=cut

=head2 remote_url( $remote_name )

Returns a URL for the remote you've requested by name.  Defaults to 'origin'.
Provides you with the exact URL which git returns. Nothing is fixed up for you.

    # defaults to 'origin'
    my $remote_url = remote_url();
    # $remote_url is now possibly something like one of the following:
    # git@github.com:oalders/git-helpers.git
    # https://github.com/oalders/git-helpers.git

    # get URL for upstream remote
    my $upstream_url = remote_url('upstream');


lib/Git/Helpers/CPAN.pm
package Git::Helpers::CPAN;
our $VERSION = '1.000002';
use Moo;

use MetaCPAN::Client 2.029000 ();
use Try::Tiny qw( try );
use Types::Standard qw( HashRef InstanceOf Maybe Str );

has _client => (
    is      => 'ro',
    isa     => InstanceOf ['MetaCPAN::Client'],
    lazy    => 1,
    default => sub { MetaCPAN::Client->new },
);

has _latest_release => (
    is      => 'ro',
    isa     => InstanceOf ['MetaCPAN::Client::Release'],
    lazy    => 1,
    builder => '_build_latest_release',
);

has release_name => (
    is      => 'ro',
    isa     => Str,
    lazy    => 1,
    builder => '_build_release_name',
);

has name => (
    is       => 'ro',
    isa      => Str,
    required => 1,
);

has repository => (
    is       => 'ro',
    isa      => Maybe [HashRef],
    lazy     => 1,
    init_arg => undef,
    builder  => '_build_repository',
);

sub _build_repository {
    my $self = shift;

    return $self->_latest_release
        ? $self->_latest_release->resources->{repository}
        : undef;
}

sub _build_release_name {
    my $self = shift;

    return $self->name if $self->name !~ m{::};

    my $release_name;
    try {
        my $module = $self->_client->module( $self->name );
        $release_name = $module->distribution if $module;
    };

    if ( !$release_name ) {
        die sprintf( "Cannot find a module named %s", $self->name );
    }

    return $release_name;
}

sub _build_latest_release {
    my $self = shift;

    my $release;
    try {
        $release = $self->_client->release( $self->release_name );
    };

    if ( !$release ) {
        die sprintf( "Cannot find a release named %s", $self->release_name );
    }
    return $release;
}

1;

# ABSTRACT: Get repository information for a CPAN module or release

=pod

=head1 SYNOPSIS

    my $by_module = Git::Helpers::CPAN->new( name => 'Git::Helpers' );
    print $by_module->repository->{web};

    my $by_release = Git::Helpers::CPAN->new( name => 'Git::Helpers' );
    print $by_release->repository->{web};

=head1 CONSTRUCTOR ARGUMENTS

=head2 name

Can be either the name of a CPAN module or a CPAN release.  In the case of this
module, you could either search for C<Git::Helpers> or C<Git-Helpers>.  For
other modules, keep in mind that the release name may not map well to the
module name.   For example C<LWP::UserAgent> and C<libwww-perl>.

=head1 METHODS

=head2 release_name

Returns a C<string> which is the actual release name the search is performed
on.  Mostly helpful for debugging.  It will match your original C<name> arg
except when you've provided a module name.  In that case this will return the
name of the release which the module maps to.

=head2 repository

Returns a C<HashRef> of repository information.  It might return something like:

    {
        type => 'git',
        url  => 'https://github.com/oalders/git-helpers.git',
        web  => 'https://github.com/oalders/git-helpers',
    }

This is essentially the data structure which is returned by the MetaCPAN API,
so it *could* change if/when the MetaCPAN API changes output formats.

This method returns C<undef> if the release was found but does not provide any
repository information.  It will C<die> if the release cannot be found via the
MetaCPAN API.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.024.
use strict;
use warnings;

use 5.012000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Shortcuts for common Git commands",
  "AUTHOR" => "Olaf Alders <olaf\@wundercounter.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Git-Helpers",
  "EXE_FILES" => [
    "script/cpan-repo",
    "script/delete-git-branches",
    "script/gh-open"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.012000",
  "NAME" => "Git::Helpers",
  "PREREQ_PM" => {
    "Browser::Open" => 0,
    "Capture::Tiny" => 0,
    "Carp" => 0,
    "File::pushd" => 0,
    "Getopt::Long" => 0,
    "Git::Sub" => 0,
    "MetaCPAN::Client" => "2.029000",
    "Moo" => 0,
    "MooX::Options" => 0,
    "Pod::Usage" => 0,
    "String::Trim" => 0,
    "Sub::Exporter" => 0,
    "Term::Choose" => "1.743",
    "Try::Tiny" => 0,
    "Types::Standard" => 0,
    "URI" => 0,
    "URI::Heuristic" => 0,
    "URI::git" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "File::Touch" => 0,
    "Git::Version" => 0,
    "Path::Tiny" => 0,
    "Test::Deep" => 0,
    "Test::Fatal" => 0,
    "Test::Git" => "1.313",
    "Test::More" => 0,
    "Test::Requires::Git" => "1.005",
    "Test::RequiresInternet" => 0
  },
  "VERSION" => "1.000002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Browser::Open" => 0,
  "Capture::Tiny" => 0,
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "File::Touch" => 0,
  "File::pushd" => 0,
  "Getopt::Long" => 0,
  "Git::Sub" => 0,
  "Git::Version" => 0,
  "MetaCPAN::Client" => "2.029000",
  "Moo" => 0,
  "MooX::Options" => 0,
  "Path::Tiny" => 0,
  "Pod::Usage" => 0,
  "String::Trim" => 0,
  "Sub::Exporter" => 0,
  "Term::Choose" => "1.743",
  "Test::Deep" => 0,
  "Test::Fatal" => 0,
  "Test::Git" => "1.313",
  "Test::More" => 0,
  "Test::Requires::Git" => "1.005",
  "Test::RequiresInternet" => 0,
  "Try::Tiny" => 0,
  "Types::Standard" => 0,
  "URI" => 0,
  "URI::Heuristic" => 0,
  "URI::git" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
